src/commands/distribute/lib/mc-fus-uploader/mc-fus-uploader.ts

```diff
@@ -0,0 +1,777 @@
+    if (this.uploadStatus.State === McFusUploadState.Verifying || this.uploadStatus.State === McFusUploadState.Completed) {
+      return;
+    }
+
+    this.setState(McFusUploadState.Verifying);
+    this.log("Verifying upload on server.");
+    const self = this;
+    this.sendRequest({
+      type: "POST",
+      useAuthentication: true,
+      url:
+        self.uploadBaseUrls.UploadFinished +
+        encodeURIComponent(self.uploadData.AssetId) +
+        "?callback=" +
+        encodeURIComponent(self.uploadData.CallbackUrl),
+      error: function (err: Error) {
+        self.log("Finalize upload failed. Trying to autorecover... " + err.message);
+        self.setState(McFusUploadState.Uploading);
+        self.startUpload();
+      },
+      success: function (response: any) {
+        // it's possible that the health check called complete before this method did.
+        // Log the current status and proceed with response verification.
+        if (self.uploadStatus.State !== McFusUploadState.Verifying) {
+          self.log("Verifying: Upload status has changed, current status: " + self.uploadStatus.State);
+        }
+
+        //if no error then execute callback
+        if (response.error === false && response.state === "Done") {
+          self.log("UploadFinalized. McFus reported the upload as completed. Status message: " + response.message, {
+            location: response.location,
+          });
+
+          // Finally report upload completion.
+          self.completeUpload();
+
+          // Attempt to perform a callback
+          self.invokeCallback(response.raw_location);
+        } else {
+          // if chunks are missing enqueue missing chunks
+          if (response.missing_chunks && response.missing_chunks.length > 0) {
+            // If there are missing chunks lets adjust the completed count.
+            self.uploadStatus.BlocksCompleted = self.uploadData.TotalBlocks - response.missing_chunks.length;
+
+            self.enqueueChunks(response.missing_chunks);
+            self.setState(McFusUploadState.Uploading);
+
+            self.log("Finalizing found missing " + response.missing_chunks.length + " chunks. Requeuing chunks.", {
+              ChunksMissing: response.missing_chunks,
+            });
+
+            const concurrentUploads = Math.min(self.maxNumberOfConcurrentUploads, self.uploadStatus.ChunkQueue.length);
+            for (let i = 0; i < concurrentUploads; i++) {
+              self.singleThreadedUpload();
+            }
+            return;
+          }
+
+          // if no chunks are missing this must be an unhandled error
+          // display the details to the user and stop the upload.
+          self.error(response.message);
+        }
+      },
+    });
+  }
+
+  hasRequiredSettings(settings: IRequiredSettings) {
+    let hasSettings = true;
+
+    if (!settings.AssetId) {
+      hasSettings = false;
+      this.error("An AssetId must be specified.");
```

Comment:
Since the error method will print this message to the user, it needs to be understandable. AssetId here is internal logic that users don't need to know about. What does the complete error log look like to the user? Is it clear that they should retry, or that something is wrong in the server side? (whichever is true)

Reply:
This is an internal library error. In CLI itself I've added a [check](https://github.com/microsoft/appcenter-cli/pull/909/files/73c25e3f2e76ef77865c0fdba1e36bd15d471c45#diff-4430e93653a0882229732bf37a3b821eR359) for this asset id and it should also log an actual error why this asset id is missing.

Reply:
So from CLI, ideally we shouldn't fall here. But the library still should have this check.